In [5]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [6]:
dataset_path="../CNN/PetImages"

In [7]:
datagen= ImageDataGenerator(
    rescale=1./255,


#Data augmentation

shear_range=0.2,
zoom_range=0.2,
horizontal_flip=True,
validation_split=0.2 

)

In [8]:
# load training data
# Train data 

train_data = datagen.flow_from_directory(
    dataset_path,
    target_size=(128, 128),
    batch_size=32,
    class_mode='binary',
    subset='training'
)

Found 20000 images belonging to 2 classes.


In [9]:
#Load validation data

val_data=datagen.flow_from_directory(
    dataset_path,
    target_size=(128,128),
    batch_size=32,
    class_mode='binary',
    subset='validation'
)

Found 4998 images belonging to 2 classes.


In [10]:
print(train_data.class_indices)

{'Cat': 0, 'Dog': 1}


In [11]:
#Building CNN model

model=keras.Sequential()

In [12]:
#First Convulational layer
#Based on the dataset and its complexity we decide how many convulational and pooling layers are required.


model.add(
    keras.layers.Conv2D(
        filters=32,
        kernel_size=(3,3),
        activation='relu',
        input_shape=(128,128,3),
    )
)

d:\Fullstackdatascientist\DL\tfenv\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [13]:
#First pooling layer

model.add(

    keras.layers.MaxPooling2D(
        pool_size=(2,2)
    ),
)

In [14]:
# 2 Conv Layer

model.add(
    keras.layers.Conv2D(
        filters=32,
        kernel_size=(3,3),
        activation='relu'
    )
)

In [15]:
# 2 Pooling Layer

model.add(

    keras.layers.MaxPooling2D(
        pool_size=(2,2)
    )
)


In [16]:
# Flatten

model.add(keras.layers.Flatten())

In [17]:
# Dense Layer

model.add(
    keras.layers.Dense(
        units=128,
        activation='relu'
    )
)

In [18]:
# Output Layer

model.add(
    keras.layers.Dense(
        units=1,
        activation='sigmoid'
    )
)

In [19]:
#view Model Architecture

model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 126, 126, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 63, 63, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 61, 61, 32)     │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 30, 30, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 28800)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │     3,686,528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,696,801 (14.10 MB)

 Trainable params: 3,696,801 (14.10 MB)

 Non-trainable params: 0 (0.00 B)

(None,126,126,Channels)

(Batch size,Height,width,No of Filters)

Input=128*128*3

filter=3*3

no of filters=32

output_size=((input-kernel)/stride)+1

stride=1

output_size=((128-3)/1)+1=125+1=126

height=126   width=126

filters=32

final_output=(None,126,126,32)

Kernel

Formula=(kernel_height*Kernel_width*Input_Channles+1)*Filters

kernels=3*3

Input_channels=3

Filters=32

params=(3 * 3 * 3 + 1) * 32
      
      =(27+1) * 32 = 896

### Pooling layer

MaxPooling((2,2))

pooling=No learning,No weights

(only reuces size)

Input=126*126*32

pool=2*2

Formula=126/2

  =63

Output:63*63*32

Final:(None,63,63,32)

### Second Convolution layer

Input from pooling : (None,63,63,32)

kernel = 3*3 

output = (63-3+1) = 61

(None,61,61,32)

params=9248

=(3 * 3 * 32 + 1) * 32

=289*32

=9248

### Second Pooling

(None,30,30,32)

Param=0

input=(61,61,32)

Pooling=2*2

Floor divison = 61/2 = 30

Final:(None,30,30,32)

No Parameters.





  

In [20]:
! pip install pillow

In [21]:
import tensorflow as tf
print(tf.config.list_physical_devices())


[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]


In [22]:
# Compile CNN
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

In [ ]:
#Evaluation
loss,accuracy=model.evaluate(val_data) 
print(f"Loss: {loss}")
print(f"Accuracy: {accuracy}")

157/157 ━━━━━━━━━━━━━━━━━━━━ 19s 115ms/step - accuracy: 0.5062 - loss: 0.6948
Loss: 0.6948125958442688
Accuracy: 0.5062024593353271


In [25]:
history=model.fit(
    train_data,
    epochs=10,
    validation_data=val_data
)

Epoch 1/10
102/625 ━━━━━━━━━━━━━━━━━━━━ 2:00 230ms/step - accuracy: 0.5353 - loss: 0.8389

d:\Fullstackdatascientist\DL\tfenv\Lib\site-packages\PIL\TiffImagePlugin.py:949: UserWarning: Truncated File Read
  warnings.warn(str(msg))


625/625 ━━━━━━━━━━━━━━━━━━━━ 161s 256ms/step - accuracy: 0.6463 - loss: 0.6357 - val_accuracy: 0.6979 - val_loss: 0.5750
Epoch 2/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 268s 429ms/step - accuracy: 0.7240 - loss: 0.5460 - val_accuracy: 0.7553 - val_loss: 0.4943
Epoch 3/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 319s 511ms/step - accuracy: 0.7642 - loss: 0.4859 - val_accuracy: 0.7797 - val_loss: 0.4657
Epoch 4/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 138s 220ms/step - accuracy: 0.7896 - loss: 0.4481 - val_accuracy: 0.7919 - val_loss: 0.4413
Epoch 5/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 123s 196ms/step - accuracy: 0.8083 - loss: 0.4179 - val_accuracy: 0.7927 - val_loss: 0.4423
Epoch 6/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 129s 206ms/step - accuracy: 0.8197 - loss: 0.3984 - val_accuracy: 0.8149 - val_loss: 0.4109
Epoch 7/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 233s 372ms/step - accuracy: 0.8317 - loss: 0.3751 - val_accuracy: 0.8219 - val_loss: 0.3988
Epoch 8/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 202s 323ms/step - accuracy: 0.8389 - loss: 0.35

In [26]:
#Evaluation
loss,accuracy=model.evaluate(val_data) 
print(f"Loss: {loss}")
print(f"Accuracy: {accuracy}") 

157/157 ━━━━━━━━━━━━━━━━━━━━ 23s 146ms/step - accuracy: 0.8305 - loss: 0.3853
Loss: 0.38529276847839355
Accuracy: 0.8305321931838989
